### ISSUE TITLE
机器学习速成课程（谷歌） 第一次讨论

### ISSUE THREAD

**Issue by zengbin93** [zengbin93/ML-Courses#4]
## 前四课内容概述

**第 1 课** 三分钟课程介绍。

**第 2 课** 介绍了监督学习的一些术语，如标签 label，特征 features，回归（输出连续值）与分类（输出离散值）。

**第 3 课** 以线性回归为例介绍了机器学习的流程：定义预测函数，得到损失，梯度下降更新参数。
线性回归的损失函数采用 Mean Square Error，Squared loss 也叫 L2 loss，L2 loss 比 L1 loss 更惩罚 outlier。模型训练的过程就是找到合适的参数使的损失不断降低的过程，这个过程叫做 **经验风险最小化 empirical risk minimization ERM**。与经验风险最小化对应的是 **结构风险最小化 Structural risk minimization SRM**，在保证模型预测能力的同时要求模型尽可能的简单，即最小化 (loss + regularization),  SRM 的提出是为了防止过拟合。李航老师的《统计学习方法》中说，最大似然估计 MLE 属于 ERM 范畴，最大后验概率 MAP 属于 SRM 范畴

**第 4 课** 介绍了模型参数求解方法-**梯度下降法**：每次朝着负梯度方向更新参数。梯度是一个向量，有大小和方向，其每个元素是对相应权重的偏导数。对于线性回归来说，loss 函数是凸函数，只有一个最小值（此点导数为 0 ），因此初始值不敏感，设置成随机值或者 0 都可以。神经网络不是凸函数，有很多局部极小值点（想象装蛋箱 egg crate）, 因此依赖于初始值。参数初始值有 He initialization, Xavier initialization 等。学习率 learning rate 是个重要的概念，它是一个超参数Hyperparameter，其值太小则收敛速度慢，其值太大则可能不收敛，通常初始值为 0.001 (个人经验，有待讨论)。通常采取的是 mini-batch gradient descent，batch 大小 n 通常为 2 的幂次方（即 16，32，64，128，256等），而当 n=1 时叫做随机梯度下降 stochastic gradient descent。Google 提供了 http://playground.tensorflow.org/ 供可视化调参数，非常有意思。

**Comment by zengbin93** [comment]
**问题一：有人说大多数机器学习算法的本质都是回归，这个说法对吗？**
---

@曾斌 我觉得回归似乎是隐含了label之间有某种结构性或顺序，但感觉分类才是最基本的机器学习任务？我记得karpathy在他那个cs231n的notes里也说分类才是更基本的

@DV 嗯嗯，分类也是很基本的任务，但我感觉很多分类算法的本质也是回归，即在空间中构建分割线

嗯，逻辑回归那种都是generalized linear model (GLM)，其实只是把回归模型换了个link function，不过tree-based那种感觉还是差别比较大@曾斌 

对，基于树结构的分类模型和回归有比较明显的区分，而且解释性往往更强
